### PR TITLE
Add CoreDNS as an optional addon in kube-up

### LINKED ISSUE #56439: Add CoreDNS in kube-up
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
> Uncomment only one, leave it on its own line: 
>
> /kind bug
>/kind feature

**What happened**:
Proceeding to add CoreDNS as an addon option in kube-up which can be used instead of kube-dns.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig network

### PR DESCRIPTION
**Pull request by rajansandeep**
**What this PR does / why we need it**:
This PR adds the option of installing CoreDNS as an addon instead of kube-dns in kube-up.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56439 

**Special notes for your reviewer**:

**Release note**:

```release-note
kube-up: Add optional addon CoreDNS. 
Install CoreDNS instead of kube-dns by setting CLUSTER_DNS_CORE_DNS value to 'true'.
```

### CODE REVIEW

**Comment by bowei** [comment]
How do you run this? You will need to add a release-note comment regarding about to use this. Please put it in the commit and the PR description above.

**Comment by roberthbailey** [comment]
I will take a look once @bowei or his delegate approves.

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -1837,22 +1837,32 @@ EOF
...
+    if [[ "${CLUSTER_DNS_CORE_DNS:-}" == "true" ]]; then
+      setup-addon-manifests "addons" "dns"
+      local -r coredns_file="${dst_dir}/dns/coredns.yaml"
+      mv "${dst_dir}/dns/coredns.yaml.in" "${coredns_file}"
+      # Replace the salt configurations with variable values.
+      sed -i -e "s@{{ *pillar\['dns_domain'\] *}}@${DNS_DOMAIN}@g" "${coredns_file}"
+      sed -i -e "s@{{ *pillar\['dns_server'\] *}}@${DNS_SERVER_IP}@g" "${coredns_file}"
+      sed -i -e "s@{{ *pillar\['service_cluster_ip_range'\] *}}@${SERVICE_CLUSTER_IP_RANGE}@g" "${coredns_file}"
+    else
+      setup-addon-manifests "addons" "dns"
```
Comment (bowei) [inline cluster/gce/gci/configure-helper.sh]: this looks to be repeated on both branches of if - else

File: cluster/gce/container-linux/configure-helper.sh [hunk 2]
```diff
@@ -1293,22 +1293,32 @@ EOF
     setup-addon-manifests "addons" "metrics-server"
   fi
   if [[ "${ENABLE_CLUSTER_DNS:-}" == "true" ]]; then
-    setup-addon-manifests "addons" "dns"
-    local -r kubedns_file="${dst_dir}/dns/kube-dns.yaml"
-    mv "${dst_dir}/dns/kube-dns.yaml.in" "${kubedns_file}"
-    if [ -n "${CUSTOM_KUBE_DNS_YAML:-}" ]; then
-      # Replace with custom GKE kube-dns deployment.
-      cat > "${kubedns_file}" <<EOF
+    if [[ "${CLUSTER_DNS_CORE_DNS:-}" == "true" ]]; then
```
Comment (bowei) [inline cluster/gce/container-linux/configure-helper.sh]: can you split this off into two bash functions

setup-coredns-manifest()
setup-kube-dns-manifest()

to make it easier to read

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -1837,22 +1837,32 @@ EOF
     setup-addon-manifests "addons" "metrics-server"
   fi
   if [[ "${ENABLE_CLUSTER_DNS:-}" == "true" ]]; then
-    setup-addon-manifests "addons" "dns"
-    local -r kubedns_file="${dst_dir}/dns/kube-dns.yaml"
-    mv "${dst_dir}/dns/kube-dns.yaml.in" "${kubedns_file}"
-    if [ -n "${CUSTOM_KUBE_DNS_YAML:-}" ]; then
-      # Replace with custom GKE kube-dns deployment.
-      cat > "${kubedns_file}" <<EOF
+    if [[ "${CLUSTER_DNS_CORE_DNS:-}" == "true" ]]; then
```
Comment (bowei) [inline cluster/gce/gci/configure-helper.sh]: see comment above

File: cluster/addons/dns/coredns.yaml.sed [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+metadata:
+  name: coredns
+  namespace: kube-system
+  labels:
+    k8s-app: coredns
+    kubernetes.io/cluster-service: "true"
+    addonmanager.kubernetes.io/mode: Reconcile
+    kubernetes.io/name: "CoreDNS"
+spec:
+  replicas: 1
```
Comment (bowei) [inline cluster/addons/dns/coredns.yaml.sed]: Should probably have 2 for redundancy

File: cluster/addons/dns/coredns.yaml.sed [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  selector:
+    matchLabels:
+      k8s-app: coredns
+  template:
+    metadata:
+      labels:
+        k8s-app: coredns
+    spec:
```
Comment (bowei) [inline cluster/addons/dns/coredns.yaml.sed]: no resource definitions?

Comment (rajansandeep) [reply cluster/addons/dns/coredns.yaml.sed]: I added them now.

File: cluster/addons/dns/coredns.yaml.sed (hunk 4 above)
Comment (rajansandeep) [reply cluster/addons/dns/coredns.yaml.sed]: The manifests are defined to mimic the ones in Kubeadm.

File: cluster/gce/container-linux/configure-helper.sh [hunk 6]
```diff
@@ -1228,6 +1228,36 @@ function update-prometheus-to-sd-parameters {
   fi
 }
 
+# Prepares the manifests of coreDNS for k8s addons.
+function setup-coredns-manifest {
+  local -r coredns_file="${dst_dir}/dns/coredns.yaml"
+      mv "${dst_dir}/dns/coredns.yaml.in" "${coredns_file}"
```
Comment (bowei) [inline cluster/gce/container-linux/configure-helper.sh]: strange indent here

File: cluster/gce/container-linux/configure-helper.sh [hunk 7]
```diff
@@ -1228,6 +1228,36 @@ function update-prometheus-to-sd-parameters {
...
+      # Replace the salt configurations with variable values.
+      sed -i -e "s@{{ *pillar\['dns_domain'\] *}}@${DNS_DOMAIN}@g" "${coredns_file}"
+      sed -i -e "s@{{ *pillar\['dns_server'\] *}}@${DNS_SERVER_IP}@g" "${coredns_file}"
+      sed -i -e "s@{{ *pillar\['service_cluster_ip_range'\] *}}@${SERVICE_CLUSTER_IP_RANGE}@g" "${coredns_file}"
+}
+
+# Prepares the manifests of kube-dns for k8s addons.
+function setup-kube-dns-manifest {
+  local -r kubedns_file="${dst_dir}/dns/kube-dns.yaml"
+      mv "${dst_dir}/dns/kube-dns.yaml.in" "${kubedns_file}"
```
Comment (bowei) [inline cluster/gce/container-linux/configure-helper.sh]: fix the indent

File: cluster/gce/container-linux/configure-helper.sh [hunk 8]
```diff
@@ -1228,6 +1228,36 @@ function update-prometheus-to-sd-parameters {
...
+      if [ -n "${CUSTOM_KUBE_DNS_YAML:-}" ]; then
+        # Replace with custom GKE kube-dns deployment.
+        cat > "${kubedns_file}" <<EOF
+$(echo "$CUSTOM_KUBE_DNS_YAML")
+EOF
+        update-prometheus-to-sd-parameters ${kubedns_file}
```
Comment (bowei) [inline cluster/gce/container-linux/configure-helper.sh]: consistent indent...

**Review by bowei — COMMENTED** [review]
looks ok

need to fix the indentation... in the sh

**Comment by bowei** [comment]
@roberthbailey you have the approval

**Comment by bowei** [comment]
@rajansandeep ping me on slack if this is running up against the code freeze deadline

**Comment by bowei** [comment]
You need to add milestone and @roberthbailey need to approve...

**Comment by rajansandeep** [comment]
@bowei how is the milestone label added?

**Comment by cblecker** [comment]
@rajansandeep @mikedanese Is there an associated issue for this (or `/approve no-issue`)? That's what it's missing to merge.

**Comment by rajansandeep** [comment]
@cblecker There is no specific issue associated as such. Needs a no issue approve label.

**Comment by bowei** [comment]
There should be an issue for the feature, can you create?

**Comment by rajansandeep** [comment]
@bowei I created the issue and edited the comment to include it.

**Pull request merged**